scripts/test-kic-2.x-upgrade.sh

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env bash
+
+# ------------------------------------------------------------------------------
+# test-kick-2.x-upgrade.sh
+#
+# This script is temporary: in the timeframe between 1.x and 2.x KIC
+# releases this script was made to validate that upgrading to the
+# latest pre-release image was functional, once 2.0 releases fully
+# this script will be removed because all tests thereafter will be
+# running on that version and it's lineage.
+# ------------------------------------------------------------------------------
+
+# ------------------------------------------------------------------------------
+# Environment Variables
+# ------------------------------------------------------------------------------
+
+if [[ -z $TEST_ENV_NAME ]]
+then
+    TEST_ENV_NAME="kong-charts-tests"
+fi
+
+# ------------------------------------------------------------------------------
+# Shell Configuration
+# ------------------------------------------------------------------------------
+
+set -euo pipefail
+
+# ------------------------------------------------------------------------------
+# Kubernetes Cluster Setup
+# ------------------------------------------------------------------------------
+
+kubectl cluster-info --context kind-${TEST_ENV_NAME} 1>/dev/null
+KUBERNETES_VERSION="$(kubectl version -o json | jq -r '.serverVersion.gitVersion')"
+
+# ------------------------------------------------------------------------------
+# Determine the latest KIC 2.x Pre-Release Version
+# ------------------------------------------------------------------------------
+
+BASE="kong/kubernetes-ingress-controller"
+LATEST_PRERELEASE=$(curl -s https://api.github.com/repos/${BASE}/releases | \
```

Comment:
Could we just run against the `next-railgun` tag instead of trying to guess that latest pre-release here?

That will give us earlier feedback, without the need to wait for the pre-release to happen.

This suggestion boils down to redefining this test as "please test the upgrade from version X to version Y" where the default X is what's set in the chart, and the default Y is `railgun-next`.

Reply:
:+1: and we may also want to consider moving this to a cron trigger rather than a push trigger: KIC will get continuous pre-release (in the "it's committed to next" sense, not in the "it's a released alpha/beta image" sense) changes independent of the chart, and testing the chart against next will call out when we make changes that the chart must account for. We probably want to observe those as they happen in KIC, rather than waiting for a (probably unrelated) push to the chart to trigger it.

For specific KIC versions, we can probably test upgrades via `ct install` tests, which support a `--upgrade` mode to compare a previous chart version to a proposed version. That proposed version can include a values.yaml with a version bump for images, and testing that upgrade will thus test an image upgrade.

`ct` version upgrade testing is static (it is always the exact image versions you have in the previous and proposed values.yaml), but that's probably what we want normally. The KIC 1.x -> 2.x transition is unique and we probably want to test that migration as a one-off thing outside, whereas CI will be concerned with more routine "default minor version upgrade succeeds" checks that we'll have ongoing going forward.

Reply:
added in ab26afcebebe82f212918915b0814616800d408b :+1: